### PR TITLE
UNR-3908-Test-TearOff-Prevents-Actor-Replication

### PR DESCRIPTION
**Pull request by AndreiTeodorLazar**
#### Description
This PR adds an automated test that checks if calling TearOff prevents Actor from replication, as it should be covered by the QA workflow C2413 (more details in the image below). However, this workflow currently does not have steps on how to test. Therefore, this PR is proposing a solution to that particular flow.

image

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+/**
+ * This test automates the QA flow that checks if Actors can be prevented from replicating with bTearOff.
+ * NOTE: Due to adding wait steps before checking certain parts of the test, this test runs slow, in roughly 30 seconds. This can be improved by modifying the WorkerWaitForTimeStepDefinition
+ *       to wait for a smaller amount of time. 
+ *
+ * The test includes 1 Server and 2 Client workers.
+ * The flow is as follows:
+ * - Setup:
+ *  - A ReplicatedTearOffActor is placed in the map that is running the test.
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:18]: This makes it sound like the test does this, but it actually requires a map with it preplaced.
On a related note, any reason it needs to be a start-up actor? Is that the QA flow?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:18]: I see it has both flows, so I would just reword this to say one needs to be preplaced in the map.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:18]: Okay this is something I should have been specific about in the PR, there is currently no 'accepted' test flow for covering  the bTearOff QA workflow, this PR is basically trying to achieve that with an automated test. So yeah I thought that the 3 main flows would be: a startup Actor that calls TearOff on BeginPlay(), a dynamically spawned Actor with the same behaviour, and a dynamically spawned 'normal'replicated Actor that has TearOff called at some point. I have just updated the PR comment to make this more clear.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:18]: Related to the wording, I agree it is confusing, I will update that, thank you for the suggestion!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+				if (IsValid(SpawnedReplicatedActorBase) && SpawnedReplicatedActorBase->GetActorLocation().Equals(SpwanedReplicatedActorBaseLocation, 1.0f))
+				{
+					FinishStep();
+				}
+			}
+		}, 5.0f);
+
+	// Tear off the spawned replicated Actor and move it to a new location.
+	AddStep(TEXT("SpatialTestTearOffServerSetTearOffAndMoveActor"), FWorkerDefinition::Server(1), nullptr, [this]()
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:168]: Can you add an extra step showing you were able to replicate position before calling tear off ?

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:168]: Will do!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	virtual void BeginPlay() override;
+
+	// Variable used to reference the ReplicatedTearOffActor placed in the map.
+	AReplicatedTearOffActor* StartupTearOffActor;
+
+	// Variable used to reference the spawned AReplicatedTestActorBase.
+	AReplicatedTestActorBase* SpawnedReplicatedActorBase;
+
+	// The location of the spawned ReplicatedActorBase.
+	FVector SpwanedReplicatedActorBaseLocation;
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h:29]:
```suggestion
	FVector SpawnedReplicatedActorBaseLocation;
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+			if (StartupTearOffActor && StartupTearOffActor->TestInteger == 0)
+			{
+				FinishStep();
+			}
+		});
+
+	// The server dynamically spawns a ReplicatedTearOffActor.
+	AddStep(TEXT("SpatialTestTearOffServerSpawnReplicatedTearOffActor"), FWorkerDefinition::Server(1), nullptr, [this]()
+		{
+			AReplicatedTearOffActor* SpawnedTearOffActor = GetWorld()->SpawnActor<AReplicatedTearOffActor>(FVector(50.0f, 50.0f, 80.0f), FRotator::ZeroRotator, FActorSpawnParameters());
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:106]: I see this is just coded here. Since there is a spawn & movement flow, I think I would slighly prefer all of these locations to be in variables and declared somewhere close together.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:106]: Sure thing, will add this!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+					FinishStep();
+				}
+			}
+		}, 5.0f);
+
+	// Tear off the spawned replicated Actor and move it to a new location.
+	AddStep(TEXT("SpatialTestTearOffServerSetTearOffAndMoveActor"), FWorkerDefinition::Server(1), nullptr, [this]()
+		{
+			SpawnedReplicatedActorBase->TearOff();
+			SpawnedReplicatedActorBase->SetActorLocation(FVector(30.0f, -20.0f, 80.0f));
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:171]: Does this not propagate in native? I would have thought that since you mentioned native has different semantics that this would pose a problem in native.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:171]: I'm not sure I am following what you mean, but the difference that I have noticed with Native is that if you have a Startup Actor that calls TearOff on BeginPlay, then it takes longer to actually prevent replication than in Spatial (did not dig too much into this to see why this happens). In this case however I didn't notice any difference.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:171]: Okay, I just thought if native takes some time to actually tear off in begin play, that the order of (TearOff) (ReplicateProperty) in the same tick would also not be guaranteed. But if it works, then it works.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:171]: Out of curiosity, I would want to take a look at the BeginPlay case, but I am not sure what's a good approach for this, can you think of a way of checking/ a starting point on  why the tear off in BeginPlay takes a bit longer?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.cpp:171]: Would be kind of annoying - would have to check the native replication flow and see where bTearOff is checked and why it would not be respected / alternatively if a replication tick happens before BeginPlay (but I believe that shouldn't be the case).

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+public:
+	ASpatialTestTearOff();
+
+	virtual void BeginPlay() override;
+
+	// Variable used to reference the ReplicatedTearOffActor placed in the map.
+	AReplicatedTearOffActor* StartupTearOffActor;
+
+	// Variable used to reference the spawned AReplicatedTestActorBase.
+	AReplicatedTestActorBase* SpawnedReplicatedActorBase;
```
Comment (padhansell) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h:26]: Since this is will be the only reference to the newly spawned actor from SpawnActor(), I think it should be marked UPROPERTY() to prevent garbage collection?

**Review by padhansell — CHANGES_REQUESTED** [review]
One small question but otherwise looks good to me 👍

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h (hunk 6 above)
Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h:26]: To be honest, I do not know much about UE's garbage collection, if that is the case, then I will definitely add that. What do you think about this @ImprobableNic @mironec ?

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestTearOff/SpatialTestTearOff.h:27]: I do not think this is necessary. Spawned actors should be referenced in the Level.

**Review by ImprobableNic — APPROVED** [review]

**Review by mironec — APPROVED** [review]

**Review by padhansell — APPROVED** [review]
